stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+        stacks_client: &StacksClient,
+        block: &NakamotoBlock,
+        current_reward_cycle: u64,
+    ) -> bool {
+        let aggregate_key = retry_with_exponential_backoff(|| {
+            stacks_client
+                .get_approved_aggregate_key(self.reward_cycle)
+                .map_err(backoff::Error::transient)
+        })
+        .unwrap_or(None);
+        if aggregate_key.is_some() {
+            // We do not enforce a block contain any transactions except the aggregate votes when it is NOT already set
+            debug!("Signer #{}: Already have an aggregate key for reward cycle {}. Skipping transaction verification...", self.signer_id, self.reward_cycle);
+            return true;
+        }
+        if let Ok(expected_transactions) =
+            self.get_expected_transactions(stacks_client, current_reward_cycle)
+        {
+            //It might be worth building a hashset of the blocks' txids and checking that against the expected transaction's txid.
+            let block_tx_hashset = block.txs.iter().map(|tx| tx.txid()).collect::<HashSet<_>>();
+            // Ensure the block contains the transactions we expect
+            let missing_transactions = expected_transactions
+                .into_iter()
+                .filter_map(|tx| {
+                    if !block_tx_hashset.contains(&tx.txid()) {
+                        debug!(
+                            "Signer #{}: expected txid {} is in the block",
+                            self.signer_id,
+                            &tx.txid()
+                        );
+                        Some(tx)
+                    } else {
+                        debug!(
+                            "Signer #{}: missing expected txid {}",
+                            self.signer_id,
+                            &tx.txid()
+                        );
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            let is_valid = missing_transactions.is_empty();
+            if !is_valid {
+                debug!("Signer #{}: Broadcasting a block rejection due to missing expected transactions...", self.signer_id);
+                let block_rejection = BlockRejection::new(
+                    block.header.signer_signature_hash(),
+                    RejectCode::MissingTransactions(missing_transactions),
+                );
+                // Submit signature result to miners to observe
+                if let Err(e) = self
+                    .stackerdb
+                    .send_message_with_retry(block_rejection.into())
+                {
+                    warn!(
+                        "Signer #{}: Failed to send block rejection to stacker-db: {e:?}",
+                        self.signer_id
+                    );
+                }
+            }
+            is_valid
+        } else {
+            // Failed to connect to the stacks node to get transactions. Cannot validate the block. Reject it.
```

Comment:
I'm not sure why we expect a broadcast-rejection to work when we just failed to query the expected transactions?

Reply:
Yeah I actually removed this in a follow on PR. Will do now though since its an easy change.